swabbie-echo/src/main/kotlin/com/netflix/spinnaker/swabbie/echo/EchoNotifier.kt

```diff
@@ -16,23 +16,78 @@
+import com.netflix.spinnaker.swabbie.model.WorkConfiguration
 import com.netflix.spinnaker.swabbie.notifications.Notifier
+import com.netflix.spinnaker.swabbie.notifications.Notifier.NotificationResult
+import com.netflix.spinnaker.swabbie.notifications.Notifier.NotificationType
+import com.netflix.spinnaker.swabbie.notifications.Notifier.NotificationSeverity
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Component
+import java.lang.UnsupportedOperationException
 
 @Component
 class EchoNotifier(
-  private val echoService: EchoService
+  private val echoService: EchoService,
+  private val retrySupport: RetrySupport
 ) : Notifier {
-  override fun notify(recipient: String, additionalContext: Map<String, Any>, messageType: String) {
-    echoService.create(
-      EchoService.Notification(
-        notificationType = Notifier.NotificationType.valueOf(messageType),
-        to = recipient.split(","),
-        severity = Notifier.NotificationSeverity.HIGH,
-        source = EchoService.Notification.Source("swabbie"),
-        templateGroup = "swabbie",
-        additionalContext = additionalContext
-      )
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val timeoutMillis: Long = 5000
+  private val maxAttempts: Int = 3
+
+  override fun notify(
+    envelope: Notifier.Envelope
+  ): NotificationResult {
+    val recipient = envelope.recipient
+    // Notifications are grouped by resource type, they all have a common notification setting.
+    val workConfiguration = envelope
+      .resources.map {
+        it.second
+      }.first()
+
+    val notificationConfig = workConfiguration.notificationConfiguration
+    notificationConfig.types.forEach { notificationType ->
+      when {
+        notificationType.equals(NotificationType.EMAIL.name, true) -> {
+          try {
+            retrySupport.retry({
+              echoService.create(
+                EchoService.Notification(
+                  notificationType = NotificationType.EMAIL,
+                  to = recipient.split(","),
+                  severity = NotificationSeverity.HIGH,
+                  source = EchoService.Notification.Source("swabbie"),
+                  templateGroup = "swabbie",
+                  additionalContext = notificationContext(recipient, envelope.resources, workConfiguration.resourceType)
+                )
+              )
+            }, maxAttempts, timeoutMillis, false)
+            return NotificationResult(recipient, NotificationType.EMAIL, success = true)
+          } catch (e: Exception) {
+            log.error("Failed to send notification", e)
+            return NotificationResult(recipient, NotificationType.EMAIL, success = false)
+          }
+        }
+      }
+    }
+
+    throw UnsupportedOperationException("Notification Type not supported in $workConfiguration")
+  }
+
+  private fun String.unCamelCase(): String =
```

Comment:
Is there a commons lang function that already does this?

Reply:
See TODO comment. I'd prefer changing resource types into something that can provide a human readable name rather than trying to normalize this way. Leaving this as-is for now. I will address this in a different refactor.